src/lib/PerceptronPredictor.cc

```diff
@@ -78,13 +76,27 @@ BranchPrediction PerceptronPredictor::predict(uint64_t address, BranchType type,
   } else if (type == BranchType::Conditional) {
     if (!prediction.taken) prediction.target = address + 4;
   }
+
+  // Store the global history for correct hashing in update() --
+  // needs to be global history and not the hashed index as hashing loses
+  // information at longer global history lengths
+  ftq_.emplace_back(prediction.taken, globalHistory_);
+
+  // speculatively update global history
+  globalHistory_ =
+      ((globalHistory_ << 1) | prediction.taken) & globalHistoryMask_;
+
   return prediction;
 }
 
 void PerceptronPredictor::update(uint64_t address, bool taken,
                                  uint64_t targetAddress, BranchType type) {
-  // Work out hash index
-  uint64_t prevGlobalHistory = btbHistory_[address];
+  // Get previous branch state and prediction from FTQ
+  bool prevPrediction = ftq_.front().first;
```

Comment:
Is this assuming that the flush order matches the additions to the history/FTQ?

Reply:
Se above reply re line 160.  This is assuming that, though the assumption is founded on code/empirical review.